pkg/server/infrastructure/datastore/mysql/mysql.go

```diff
@@ -47,40 +64,249 @@ func New(ctx context.Context, cfg datastore.Config) (datastore.DataStore, error)
 
 // Add add data model
 func (m *mysql) Add(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+	entity.SetCreateTime(time.Now())
+	entity.SetUpdateTime(time.Now())
+
+	if dbAdd := m.client.Create(entity); dbAdd.Error != nil {
+		if match := errors.Is(dbAdd.Error, gorm.ErrDuplicatedKey); match {
+			return datastore.ErrRecordExist
+		}
+		return datastore.NewDBError(dbAdd.Error)
+	}
+	return nil
 }
 
 // BatchAdd batch add entity, this operation has some atomicity.
 func (m *mysql) BatchAdd(ctx context.Context, entities []datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	notRollback := make(map[string]int)
+	for i, saveEntity := range entities {
+		if err := m.Add(ctx, saveEntity); err != nil {
+			if errors.Is(err, datastore.ErrRecordExist) {
+				notRollback[saveEntity.PrimaryKey()] = 1
```

Comment:
Magic number 1 here. Consider make notRollback to map[string]bool. Fixing other implementation is an option that can be done in another PR.

Reply:
Yeah, I will change this in this PR and consider opening another PR for the other drivers.

Reply:
Done !